libwnbd/wnbd_ioctl.cpp

```diff
@@ -133,6 +137,245 @@ DWORD WnbdOpenAdapterCMDeviceInstance(PDEVINST DeviceInstance)
+        GLA = GetLastError();
+        goto exit;
+    }
+
+    InstallParams.FlagsEx |= DI_FLAGSEX_INSTALLEDDRIVER | DI_FLAGSEX_ALLOWEXCLUDEDDRVS;
+    if (!SetupDiSetDeviceInstallParamsA(DevHandle, DevInfoData, &InstallParams)) {
+        GLA = GetLastError();
+        goto exit;
+    }
+
+    if (!SetupDiBuildDriverInfoList(DevHandle, DevInfoData, SPDIT_CLASSDRIVER)) {
+        GLA = GetLastError();
+        goto exit;
+    }
+    Cleanup = TRUE;
+
+    for (DWORD index = 0;
+        SetupDiEnumDriverInfoA(DevHandle, DevInfoData, SPDIT_CLASSDRIVER, index, &DrvInfoData);
+        index++) {
+
+        if (!SetupDiGetDriverInfoDetailA(DevHandle, DevInfoData, &DrvInfoData,
+                                         DrvDetailData, DrvDetailData->cbSize, 0)
+            && ERROR_INSUFFICIENT_BUFFER != GetLastError()) {
+            continue;
+        }
+
+        Found = TRUE;
+        break;
+    }
+    if (!Found) {
+        GLA = GetLastError();
+        goto exit;
+    }
+
+
+exit:
+    if (ERROR_SUCCESS != GLA) {
+        LogError(
+            "Could not get installed driver information data. "
+            "Error: %d. Error message: %s", GLA, win32_strerror(GLA).c_str());
+    }
+
+    if (Cleanup) {
+        SetupDiDestroyDriverInfoList(DevHandle, DevInfoData, SPDIT_COMPATDRIVER);
+    }
+
+    return GLA;
+}
+
+DWORD WnbdRemoveDriver(HDEVINFO DevHandle, PSP_DEVINFO_DATA DevInfoData, PBOOL RebootRequired)
+{
+    SP_DRVINFO_DETAIL_DATA_A DrvDetailData = { 0 };
+    DrvDetailData.cbSize = sizeof DrvDetailData;
+    DWORD GLA = ERROR_SUCCESS;
+
+    GLA = GetDrvDetailData(DevHandle, DevInfoData, &DrvDetailData);
+    if (ERROR_SUCCESS != GLA) {
+        return GLA;
+    }
+
+    std::string SearchString(DrvDetailData.InfFileName);
+
+    if (!DiUninstallDevice(0, DevHandle, DevInfoData, 0, RebootRequired)) {
+        *RebootRequired = TRUE;
+    }
+
+    if (SearchString.length()) {
+        SetupUninstallOEMInfA(
+            SearchString.substr(SearchString.find_last_of("\\") + 1).c_str(), SUOI_FORCEDELETE, 0);
+    }
+
+    return GLA;
+}
+
+
+static DWORD FindDriver(HDEVINFO* DevHandle, SP_DEVINFO_DATA* DevInfoData)
+{
+    CHAR TempBuf[2048];
+    BOOL Found = FALSE;
+    DWORD GLA = ERROR_FILE_NOT_FOUND;
+
+    *DevHandle = SetupDiGetClassDevsA(0, 0, 0, DIGCF_ALLCLASSES | DIGCF_PRESENT);
+    if (INVALID_HANDLE_VALUE == *DevHandle) {
+        GLA = GetLastError();
+        goto exit;
+    }
+
+    for (DWORD I = 0; !Found && SetupDiEnumDeviceInfo(*DevHandle, I, DevInfoData); I++)
+    {
+        if (!SetupDiGetDeviceRegistryPropertyA(*DevHandle, DevInfoData, SPDRP_HARDWAREID, 0,
+            (PBYTE)TempBuf, 2046, 0)) {
+            continue;
+        }
+
+        std::string TempString(TempBuf);
+        if (TempString.find(WNBD_HARDWAREID) != std::string::npos) {
+            Found = TRUE;
+        }
+    }
+    if (!Found) {
+        goto exit;
+    }
+
+    GLA = ERROR_SUCCESS;
+
+exit:
```

Comment:
I think we should avoid unnecessary ``goto``s. If the exit label does nothing other than returning a value, we should just use ``return`` instead of ``goto``.